chips/rp2040/src/pio.rs

```diff
@@ -0,0 +1,1233 @@
+        self.set_pins_out(sm_number, pin, 1, true);
+        self.set_set_pins(sm_number, pin, 1);
+        self.sm_init(sm_number);
+        self.sm_set_enabled(sm_number, true);
+    }
+
+    pub fn blink_program_init(
+        &mut self,
+        pio_number: PIONumber,
+        sm_number: SMNumber,
+        pin: u32,
+        config: &StateMachineConfiguration,
+    ) {
+        self.sm_config(sm_number, config);
+        self.pio_number = pio_number;
+        self.gpio_init(&RPGpioPin::new(RPGpio::from_u32(pin)));
+        self.sm_set_enabled(sm_number, false);
+        self.set_pins_out(sm_number, pin, 1, true);
+        self.set_set_pins(sm_number, pin, 1);
+        self.sm_init(sm_number);
+        self.sm_set_enabled(sm_number, true);
+    }
+
+    pub fn sideset_program_init(
+        &mut self,
+        pio_number: PIONumber,
+        sm_number: SMNumber,
+        pin: u32,
+        config: &StateMachineConfiguration,
+    ) {
+        self.sm_config(sm_number, config);
+        self.pio_number = pio_number;
+        self.gpio_init(&RPGpioPin::new(RPGpio::from_u32(pin)));
+        self.gpio_init(&RPGpioPin::new(RPGpio::GPIO7));
+        self.sm_set_enabled(sm_number, false);
+        self.set_pins_out(sm_number, pin, 1, true);
+        self.set_pins_out(sm_number, 7, 1, true);
+        self.set_set_pins(sm_number, pin, 1);
+        self.set_side_set_pins(sm_number, 7);
+        self.sm_init(sm_number);
+        self.sm_set_enabled(sm_number, true);
+    }
+
+    pub fn hello_program_init(
+        &mut self,
+        pio_number: PIONumber,
+        sm_number: SMNumber,
+        pin1: u32,
+        pin2: u32,
+        config: &StateMachineConfiguration,
+    ) {
+        self.sm_config(sm_number, config);
+        self.pio_number = pio_number;
+        self.gpio_init(&RPGpioPin::new(RPGpio::from_u32(pin1)));
+        self.gpio_init(&RPGpioPin::new(RPGpio::from_u32(pin2)));
+        self.sm_set_enabled(sm_number, false);
+        self.set_pins_out(sm_number, pin1, 1, true);
+        self.set_pins_out(sm_number, pin2, 1, true);
+        self.sm_init(sm_number);
+        self.sm_set_enabled(sm_number, true);
+        self.sm_put_blocking(sm_number, 0b11000000);
+        self.sm_put_blocking(sm_number, 0);
+    }
+
+    pub fn pwm_program_init(
+        &mut self,
+        pio_number: PIONumber,
+        sm_number: SMNumber,
+        pin: u32,
+        pwm_period: u32,
+        config: &StateMachineConfiguration,
+    ) {
+        self.sm_config(sm_number, config);
+        self.pio_number = pio_number;
+        self.gpio_init(&RPGpioPin::new(RPGpio::from_u32(pin)));
+        self.sm_set_enabled(sm_number, false);
+        self.set_pins_out(sm_number, pin, 1, true);
+        self.set_side_set_pins(sm_number, pin);
+        self.sm_init(sm_number);
+        self.sm_put_blocking(sm_number, pwm_period);
+        self.sm_exec(sm_number, 0x8080_u32); // pull
+        self.sm_exec(sm_number, 0x60c0_u32); // out isr, 32
```

Comment:
same here -- named constants would be great

Reply:
Fixed, same as before, added constants and comments explaining the exact PIO commands to make searching easier.